### ISSUE TITLE
20-ICLR(reviewing)-ELECTRA: PRE-TRAINING TEXT ENCODERS AS DISCRIMINATORS RATHER THAN GENERATORS

### ISSUE THREAD

**Issue by wangqiangneu** [wangqiangneu/MT-PaperReading#25]
## 简介
用`discriminator`代替`BERT`中的`MLM generator`. 具体方法是，用一个`small MLM generator`把输入x改造成$\hat{x}$ (随机mask 15%，用这个generator生成masked token，有可能恢复原样，有可能跟原始输入不同)。用`discriminator`去二分类$\hat{x}$中的`所有`token，是不是原始的输入。训练时`generator`和`discriminator`联合学习（word embedding是shared，其他参数不共享）。然后finetuning时，只使用`discriminator`，不用`generator`了。
* 有意思的点
  - 训练效率高。BERT每次只对masked 15%的token优化，而`ELECTRA`每次从所有token学习
  - 尽管每个training step里`ELECTRA`的FLOPS高于`BERT`（因为还需要一个generator去构造$\hat{x}$），但是在相同FLOPS条件下`ELECTRA`的效果要好很多（可以理解呀，`ELECTRA`相当于每个batch的samples更多，gradient应该是更稳定的，学的也就更快了，怀疑`ELECTRA`是不是可以用更大的lr去训）
   - `ELECTRA`本质是跟CBOW with negative sampling很像，都是根据上下文推断中心词/masked token。但是用二分类正负样本去学，而不是直接学generate
   - `ELECTRA`虽然形式上像`GAN`，都有`generator`和`discriminator`，但是这里头`generator`是MLE训的，不需要`discriminator`回传gradient，好训
   - 文章中大部分报的是`dev set`上的结果，提到finetuning的时候对`seed`很敏感，方差在`0.4 GLUE`左右，因此很有必要多跑几次取平均。而`test set`上的结果，是那个leading borad上，在测试集上只跑一次，而且用了很多刷分的tricks(e.g. `ensemble`，以及针对具体任务的调整)

## 论文信息
* Author: Stanford
* [Paper](https://openreview.net/pdf?id=r1xMH1BtvB)

## 总结
* 基于`discriminative`的pre-training目测要大火了
* NMT里考虑增加一个`discriminator`，但是不要`sentence-level`的判断，应该粒度更小一点
